broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java

```diff
@@ -122,6 +134,12 @@ public void notifyMessageArriving(final String topic, final int queueId, final l
     public void notifyMessageArriving(final String topic, final int queueId, final long maxOffset, final Long tagsCode,
         long msgStoreTime, byte[] filterBitMap, Map<String, String> properties) {
         String key = this.buildKey(topic, queueId);
+
+        // record the maxOffset if notify from message store
+        if (msgStoreTime > 0) {
+            this.lastNotifyMaxOffsetCache.put(key, maxOffset);
```

Comment:
Should we examine whether the incoming maxOffset exceeds the value of the cache?

Reply:
> Should we examine whether the incoming maxOffset exceeds the value of the cache?

You mean the cached offset may be larger than the incoming one? The ReputMessageService executes in the order of commitlog offset, so the notifies would arrive in order of queue logic offset.